modules/pam_chksshpwd/pam_chksshpwd.8
'\" t
.\"     Title: pam_chksshpwd
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 01/25/2017
.\"    Manual: Linux-PAM Manual
.\"    Source: Linux-PAM Manual
.\"  Language: English
.\"
.TH "PAM_CHKSSHPWD" "8" "01/25/2017" "Linux-PAM Manual" "Linux\-PAM Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pam_chksshpwd \- Checks to see if SSH is available using the default password
.SH "SYNOPSIS"
.HP \w'\fBpam_chksshpwd\&.so\fR\ 'u
\fBpam_chksshpwd\&.so\fR
.SH "DESCRIPTION"
.PP
pam_chksshpwd is a PAM module that checks to see if SSH is enabled, and if so checks to see if the default password is unchanged\&. If SSH is enabled with password authentication, and the default password for the "pi" user is still "raspberry", the file 
/var/lib/chksshpwd/sshwarn
is created. If SSH logins are not possible with the default user account and password, this file is deleted if it exists\&.
.SH "MODULE TYPES PROVIDED"
.PP
Only the
\fBsession\fR
module type is provided\&.
.SH "RETURN VALUES"
.PP
PAM_IGNORE
.RS 4
This is the only return value of this module\&.
.RE
.SH "EXAMPLES"
.PP
The suggested usage for
/etc/pam\&.d/common-session
is:
.sp
.if n \{\
.RS 4
.\}
.nf
session  optional  pam_chksshpwd\&.so
      
.fi
.if n \{\
.RE
.\}
.sp
.SH "SEE ALSO"
.PP
\fBpam.conf\fR(5),
\fBpam.d\fR(5),
\fBpam\fR(8)
.SH "AUTHOR"
.PP
pam_chksshpwd was written by Simon Long <simon@raspberrypi\&.org>\&.


debian/changelog
pam (1.1.8-3.1+deb8u2+rpi1) jessie; urgency=medium

  * New pam_chksshpwd module added to enable warnings if SSH is insecure

 -- Simon Long <simon@raspberrypi.org>  Wed, 25 Jan 2017 17:24:56 +0000

pam (1.1.8-3.1+deb8u2) jessie; urgency=medium

  * Non-maintainer upload.
  * debian/patches-applied/pam-loginuid-in-containers: Updated with the version
    from Ubuntu, this should fix logins in containers (Closes: #726661)

 -- Evgeni Golov <evgeni@debian.org>  Sat, 12 Nov 2016 08:38:19 +0100

pam (1.1.8-3.1+deb8u1) jessie; urgency=medium

  * Non-maintainer upload.
  * Fix CVE-2015-3238: DoS/user enumeration due to blocking pipe in pam_unix
    module (Closes: #789986)

 -- Tianon Gravi <tianon@debian.org>  Wed, 06 Jan 2016 17:25:53 -0800

pam (1.1.8-3.1) unstable; urgency=high

  * Non-maintainer upload by the Security Team.
  * Fix CVE-2013-7041: case-insensitive comparison used for verifying
    passwords in the pam_userdb module (closes: #731368).
  * Fix CVE-2014-2583: multiple directory traversal issues in the
    pam_timestamp module (closes: 757555)

 -- Michael Gilbert <mgilbert@debian.org>  Sat, 09 Aug 2014 09:50:42 +0000

pam (1.1.8-3) unstable; urgency=low

  * debian/rules: On hurd, link libpam explicitly with -lpthread since glibc
    will not dynamically switch between the libc stubs and the libpthread
    implementations on this architecture.  Thanks to Samuel Thibault for the
    patch.  Closes: #743891.

 -- Steve Langasek <vorlon@debian.org>  Mon, 07 Apr 2014 17:49:38 -0700

pam (1.1.8-2) unstable; urgency=medium

  * Mark the libaudit-dev build-dependency linux-any, since it's not
    available on non-Linux archs.  Closes: #737035.

 -- Steve Langasek <vorlon@debian.org>  Thu, 13 Feb 2014 15:02:00 -0800

pam (1.1.8-1) unstable; urgency=medium

  * New upstream release.
    - includes upstream changes to pam_exec.  Closes: #670147.
    - adds support for newer hashing algorithms to pam_userdb.
      Closes: #671740.
    - fixes handling of 'quiet' argument to pam_listfile, to match the
      documentation.  Closes: #592219.
    - fixes handling of @users@@hosts netgroup syntax in access.conf.
      Closes: #681223.
    - fixes installation of the /etc/security/namespace.d directory.
      Closes: #710998.
    - 027_pam_limits_better_init_allow_explicit_root: support for reading
      /proc/1/limits is upstream, this patch now only handles the policy
      of resetting limits by default and not applying glob limits to root.
    - debian/patches/fix-manpage-crud: drop, manpages now being generated
      upstream with a newer, fixed xsltproc.
    - debian/patches/pam_env-fix-overflow.patch, pam_env-fix-dos.patch,
      glibc-2_16-compilation-fix.patch, sys-types-include.patch: drop,
      included upstream.
  * Add build-dependency on pkg-config.
  * Ensure autogenerated files are after source files in all relevant patches,
    so that regenerating documentation doesn't cause build skew.
  * Drop the --disable-regenerate-docu argument, restoring the HTML manuals
    to the libpam-doc package.  Closes: #700485.
  * No need to override dh_compress in debian/rules, it already handles .html
    files correctly.
  * debian/libpam-cracklib.prerm: use $DPKG_MAINTSCRIPT_PACKAGE_COUNT to avoid
    prematurely removing the PAM config when the package is installed for
    multiple architectures.  Closes: #647428.

 -- Steve Langasek <vorlon@debian.org>  Thu, 16 Jan 2014 00:38:42 +0000

pam (1.1.3-11) unstable; urgency=low

  [ Wookey ]
  * Disable libaudit for stage1 bootstrap.

  [ Steve Langasek ]
  * debian/patches-applied/pam-loginuid-in-containers: pam_loginuid:
    Ignore failure in user namespaces.
  * Use [linux-any] in build-deps, instead of hard-coding a list of
    non-Linux archs.  Closes: #634516.

 -- Steve Langasek <vorlon@debian.org>  Tue, 14 Jan 2014 03:33:31 +0000

pam (1.1.3-10) unstable; urgency=low

  * Fix pam-auth-update handling of trailing blank lines in the fields of
    profiles.  LP: #1160288.
  * Reintroduce libaudit support now that libaudit has been multiarched.
    Closes: #699159.

 -- Steve Langasek <vorlon@debian.org>  Sun, 20 Oct 2013 15:30:46 -0700

pam (1.1.3-9) unstable; urgency=low

  * Revert libaudit support for now, because libaudit isn't multiarched yet
    in unstable so this regresses cross-installability.  Reopens bug
    #699159.
  * Add an or'ed dependency on cdebconf, which also implements the
    xloadtemplatefile extension that prevents us from depending on just
    'debconf-2.0'.  Thanks to Régis Boudin <regis@boudin.name> for the info.
    Closes: #677278.

 -- Steve Langasek <vorlon@debian.org>  Tue, 12 Feb 2013 23:06:30 +0000

pam (1.1.3-8) unstable; urgency=low

  * Confirm NMU for bug #611136; thanks to Michael Gilbert.
    - As a side effect, there will no longer be errors from reading the
      .pam_environment twice since we are now reading it 0 times.
      LP: #955032.
  * Adjust the pam_env documentation to match the module behavior resulting
    from the previous security upload.  Closes: #693995.
  * debian/rules: never regenerate manpages at build time; this may cause
    build skew that breaks the world in a multiarch context.  LP: #1095887.
  * debian/patches-applied/glibc-2_16-compilation-fix.patch: fix missing
    include causing build failure with eglibc 2.16.  Thanks to Daniel
    Schepler <dschepler@gmail.com>.  Closes: #693450.
  * Ditch autoconf patch in favor of a build-dependency on dh-autoreconf,
    which will let us keep up-to-date with newer autotools.  In the present
    instance, this gets us aarch64 support.
  * Install pam_timestamp_check - and while we're at it, move the manpage
    to the correct binary package.  Closes: #648695.
  * Update lintian overrides to suppress some noise about hardening and
    manpages.
  * Enable audit support, by popular demand.  This should have no major
    impact unless you're also running auditd; but I reserve the right to

modules/pam_chksshpwd/pam_chksshpwd.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <crypt.h>
#include <unistd.h>
#include <shadow.h>
#include <regex.h>
#include <sys/stat.h>

/* Define which PAM interfaces we provide */
#define PAM_SM_SESSION

/* Include PAM headers */
#include <security/_pam_macros.h>
#include <security/pam_ext.h>
#include <security/pam_appl.h>
#include <security/pam_modules.h>

#define SSHWARN "/run/sshwarn"
  
/* PAM entry point for session creation */
PAM_EXTERN int pam_sm_open_session (pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	struct stat sbuf;
	struct spwd *sp;
	FILE *fp;
	char *linebuf = NULL;
	size_t nchars;
	regex_t pwregex;

	// default is no warning, so delete the flag file if it exists
	if (stat (SSHWARN, &sbuf) == 0) unlink (SSHWARN);

	// is SSH enabled?
	if (stat ("/run/sshd.pid", &sbuf) == -1) return PAM_IGNORE;

	// is password authentication for SSH enabled?
	fp = fopen ("/etc/ssh/sshd_config", "r");
	if (fp)
	{
		regcomp (&pwregex, "^PasswordAuthentication\\s*no", REG_EXTENDED);
		while (getline (&linebuf, &nchars, fp) != -1)
		{
			if (!regexec (&pwregex, linebuf, 0, NULL, 0))
			{
				free (linebuf);
				fclose (fp);
				return PAM_IGNORE;
			}
		}
		free (linebuf);
		fclose (fp);
	}

	// get the pi user entry from the shadow file
	setspent ();
	sp = getspnam ("pi");
	endspent ();

	if (sp && sp->sp_pwdp)
	{
		// there is a properly-formatted entry in the shadow file - check the password
		char *enc = crypt ("raspberry", sp->sp_pwdp);

		if (enc && !strcmp (sp->sp_pwdp, enc))
		{
			// password match - create the flag file
			fp = fopen (SSHWARN, "wb");
			fclose (fp);
		}
	}

	return PAM_IGNORE;
}

/* PAM entry point for session cleanup */
PAM_EXTERN int pam_sm_close_session (pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	return PAM_IGNORE;
}

#ifdef PAM_STATIC

struct pam_module _pam_chksshpwd_modstruct =
{
	"pam_chksshpwd",
	NULL,
	NULL,
	NULL,
	pam_sm_open_session,
	pam_sm_close_session,
	NULL,
};

#endif
